### PR TITLE
Handle partition count decrease, reduce number of errors printed

### PR DESCRIPTION
**Pull request by dvladi77**
This is to start a discussion on how we can handle PARTCNT log message from librdkafka

### CODE REVIEW

File: nmsg/kafkaio.c [hunk 1]
```diff
@@ -215,6 +230,19 @@ _kafka_set_state(kafka_ctx_t ctx, const char *func, kafka_state state)
 	ctx->state = state;
 }
 
+static void
+_kafka_set_connection_state(kafka_ctx_t ctx, const char *func, bool state)
```
Comment (shw700) [inline nmsg/kafkaio.c:234]: For legibility, I would recommend making this:
```
static void
_kafka_set_connected(kafka_ctx_t ctx, const char *func, bool connected)
```

File: nmsg/kafkaio.c [hunk 2]
```diff
@@ -215,6 +230,19 @@ _kafka_set_state(kafka_ctx_t ctx, const char *func, kafka_state state)
...
+{
+	_nmsg_dprintf(2, "%s changing status from %s to %s\n", func,
+		_kafka_connection_state_to_str(ctx->connected), _kafka_connection_state_to_str(state));
+	if (state) {
+		if (ctx->timedout > 0)
+			_nmsg_dprintf(3, "%s number of timedout messages %ld (%ld)\n", func, ctx->timedout, ctx->timedout_total);
```
Comment (shw700) [inline nmsg/kafkaio.c:240]: "%s: ..."

File: nmsg/kafkaio.c [hunk 3]
```diff
@@ -587,31 +616,55 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
 		case RD_KAFKA_RESP_ERR__ALL_BROKERS_DOWN:
+			if (ctx->connected)
+				_kafka_set_connection_state(ctx, __func__, false);
+			break;
 		case RD_KAFKA_RESP_ERR__MSG_TIMED_OUT:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
+			ctx->timedout++;
+			ctx->timedout_total++;
+			if (ctx->timedout > 0 && (ctx->timedout % 100) == 0)
+				_nmsg_dprintf(2,"%s: %ld messages timedout\n", __func__, ctx->timedout);
```
Comment (shw700) [inline nmsg/kafkaio.c:633]: s/timedout/timed out"

File: nmsg/kafkaio.c [hunk 4]
```diff
@@ -741,9 +794,14 @@ kafka_read_start(kafka_ctx_t ctx, uint8_t **buf, size_t *len)
...
 	if (ctx->message != NULL) {
 		if (ctx->message->err == RD_KAFKA_RESP_ERR_NO_ERROR) {
+			if (!ctx->connected)
+				_kafka_set_connection_state(ctx, __func__, true);
+
 			*buf = ctx->message->payload;
 			*len = ctx->message->len;
 			ctx->consumed++;
+			if ((ctx->consumed % 10000) == 0)
+				_nmsg_dprintf(4, "Consumed %ld\n",ctx->consumed);
```
Comment (shw700) [inline nmsg/kafkaio.c:804]:
```
_nmsg_dprintf(4, "(%p) Delivered %ld messages\n", ctx, ctx->delivered);
```

File: nmsg/kafkaio.c [hunk 5]
```diff
@@ -587,31 +616,55 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
-	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
-	if (ctx == NULL) {
-		_nmsg_dprintf(2, "%s: unexpected Kafka opaque is NULL", __func__);
-		return;
-	}
-	switch (err_kafka) {
+static int
+_kafka_handle_error(rd_kafka_t *rk, kafka_ctx_t ctx, rd_kafka_resp_err_t err) {
+	int level = 2;
+	switch (err) {
```
Comment (shw700) [inline nmsg/kafkaio.c:622]: new line after declarations

File: nmsg/kafkaio.c [hunk 6]
```diff
@@ -587,31 +616,55 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
+	return level;
+}
+
+static void
+_kafka_error_cb(rd_kafka_t *rk, int err, const char *reason, void *opaque)
+{
+	int level;
+	kafka_ctx_t ctx = (kafka_ctx_t) opaque;
+	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
+	if (ctx == NULL) {
```
Comment (shw700) [inline nmsg/kafkaio.c:660]: new line after declarations

File: nmsg/kafkaio.c [hunk 7]
```diff
@@ -629,19 +682,19 @@ _kafka_delivery_cb(rd_kafka_t *rk, const rd_kafka_message_t *rkmessage, void *op
...
-			ctx->dropped++;
-			level = 4;
-		}
-		_nmsg_dprintf(level, "%s: got Kafka error %d: %s\n", __func__, rkmessage->err,
-			      rd_kafka_err2str(rkmessage->err));
-
-	} else
+		int level;
+		ctx->dropped++;
+		level = _kafka_handle_error(rk, ctx, rkmessage->err);
```
Comment (shw700) [inline nmsg/kafkaio.c:687]: new line after declarations

File: nmsg/kafkaio.c (hunk 1 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Fixed

File: nmsg/kafkaio.c (hunk 2 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Fixed

File: nmsg/kafkaio.c (hunk 4 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Fixed

File: nmsg/kafkaio.c (hunk 5 above)
Comment (dvladi77) [reply nmsg/kafkaio.c:623]: Fixed

File: nmsg/kafkaio.c (hunk 6 above)
Comment (dvladi77) [reply nmsg/kafkaio.c:662]: Fixed

File: nmsg/kafkaio.c (hunk 7 above)
Comment (dvladi77) [reply nmsg/kafkaio.c:691]: Fixed

File: nmsg/kafkaio.c [hunk 8]
```diff
@@ -587,31 +616,57 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
 		case RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE:
-		/* At the moment treat any broker's error as fatal */
 		default:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
-			_kafka_set_state(ctx, __func__, kafka_state_break);
 			break;
 	}
+
+	if (ctx->last_error != err) {
+		if (ctx->last_error > 1)
```
Comment (cmikk) [inline nmsg/kafkaio.c:646]: Should this be:
```
if (ctx->last_error_count > 1)
```

File: nmsg/kafkaio.c (hunk 1 above)
Comment (shw700) [reply nmsg/kafkaio.c]: I am not seeing any change

File: nmsg/kafkaio.c (hunk 4 above)
Comment (shw700) [reply nmsg/kafkaio.c]: Note these two lines:

```
_nmsg_dprintf(6, "%s: Delivered %ld\n",__func__, ctx->delivered);
_nmsg_dprintf(6, "%s: Consumed %ld\n",__func__, ctx->consumed);
```

1. There should be a space before `__func__`
2. Delivered/consumed what? I added "messages" in suggested fix to make the output more meaningful...

Comment (shw700) [reply nmsg/kafkaio.c]: OK I see #1 fixed but not #2

File: nmsg/kafkaio.c (hunk 1 above)
Comment (shw700) [reply nmsg/kafkaio.c]: I think you are commenting on the wrong thread. I already marked this resolved

File: nmsg/kafkaio.c (hunk 4 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Missed message format change. Fixed

File: nmsg/kafkaio.c [hunk 9]
```diff
@@ -587,31 +616,57 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
-		/* At the moment treat any broker's error as fatal */
 		default:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
-			_kafka_set_state(ctx, __func__, kafka_state_break);
 			break;
 	}
+
+	if (ctx->last_error != err) {
+		if (ctx->last_error_count > 1)
+			_nmsg_dprintf(level, "%s: Suppressed %ld errors %d\n", __func__, ctx->last_error_count, ctx->last_error);
```
Comment (shw700) [inline nmsg/kafkaio.c:647]: I think `last_error_count` errors were received but `last_error_count - 1` errors were suppressed

File: nmsg/kafkaio.c [hunk 10]
```diff
@@ -587,31 +616,57 @@ _kafka_ctx_destroy(kafka_ctx_t ctx)
...
 		case RD_KAFKA_RESP_ERR__TRANSPORT:
 		case RD_KAFKA_RESP_ERR__ALL_BROKERS_DOWN:
+			if (ctx->connected)
+				_kafka_set_connected(ctx, __func__, false);
+			break;
 		case RD_KAFKA_RESP_ERR__MSG_TIMED_OUT:
-			_nmsg_dprintf(2, "%s: got Kafka error %d: %s\n", __func__, err, reason);
+			ctx->timedout++;
+			ctx->timedout_total++;
+			if (ctx->timedout > 0 && (ctx->timedout % 100) == 0)
```
Comment (shw700) [inline nmsg/kafkaio.c:633]: First conditional is unnecessary; `ctx->timedout` will always be `> 0`

File: nmsg/kafkaio.c (hunk 9 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Fixed

File: nmsg/kafkaio.c (hunk 10 above)
Comment (dvladi77) [reply nmsg/kafkaio.c]: Fixed

**Review by shw700 — APPROVED** [review]

**Review by cmikk — CHANGES_REQUESTED** [review]
I've requested some clarification changes to the logging revamp.  If you anticipate these taking some time, please consider submitting the core fix as a separate PR so we can roll that out in a timely fashion.

**Review by cmikk — APPROVED** [review]

**Pull request closed without merging**